tests/system/src/test/java/test/robot/javafx/scene/TextCharacterIndexTest.java

```diff
@@ -0,0 +1,456 @@
+ *    to the Text node.
+ *
+ * Steps for testTwoEmojiNodesCharIndexEmbeddedInTexFlow()
+ * 1. Create a TextFlow. Add two Text nodes containing only emojis.
+ * 2. Move the cursor from first character to last character with a random
+ *    incement generated in step() method.
+ * 3. Character index should change as expected and it is relative
+ *    to the Text node.
+ *
+ * Steps for testWrappedTextNodesCharIndexEmbeddedInTexFlow()
+ * 1. Create a TextFlow. Add two Text nodes which gets wrapped.
+ * 2. Move the cursor from first character to last character on both lines
+ *    with a random incement generated in step() method.
+ * 3. Character index should change as expected and it is relative
+ *    to the Text node.
+ *
+ * Steps for testMultiLineTextNodesCharIndexEmbeddedInTexFlow()
+ * 1. Create a TextFlow. Add two Text nodes with emojis where new line is explicitly added.
+ * 2. Move the cursor from first character to last character on both lines
+ *    with a random incement generated in step() method.
+ * 3. Character index should change as expected and it is relative
+ *    to the Text node.
+ *
+ * Steps for testMultiLineEmojisCharIndexEmbeddedInTexFlow()
+ * 1. Create a TextFlow. Add two Text nodes with emojis where new line is explicitly added.
+ * 2. Move the cursor from first character to last character on both lines
+ *    with a random incement generated in step() method.
+ * 3. Character index should change as expected and it is relative
+ *    to the Text node.
+ */
+
+public class TextCharacterIndexTest {
+    static CountDownLatch startupLatch = new CountDownLatch(1);
+    static Random random;
+    static Robot robot;
+    static TextFlow textFlow;
+    static Text textOne;
+    static Text textTwo;
+    static Text emoji;
+
+    static volatile Stage stage;
+    static volatile Scene scene;
+
+    static final int WIDTH = 500;
+    static final int HEIGHT = 150;
+
+    final int Y_OFFSET = 25;
+    final int X_LEADING_OFFSET = 10;
+    final int X_TRAILING_OFFSET = 40;
+
+    boolean isLeading;
+    boolean isSurrogatePair;
+    int charIndex;
+    int insertionIndex;
+    int textFlowCharIndex;
+
+    private void mouseClick(double x, double y) {
+        Util.runAndWait(() -> {
+            robot.mouseMove((int) (scene.getWindow().getX() + scene.getX() + x),
+                    (int) (scene.getWindow().getY() + scene.getY() + y));
+            robot.mouseClick(MouseButton.PRIMARY);
+        });
+    }
+
+    private void moveMouseOverTextFlow(double x, double y) throws Exception {
+        mouseClick(textFlow.getLayoutX() + x,
+                    textFlow.getLayoutY() + y);
+    }
+
+    private void addTwoTextNodes() {
+        Util.runAndWait(() -> {
+            textOne = new Text("This is Text");
+            textOne.setFont(new Font(48));
+            textTwo = new Text("This is Text");
+            textTwo.setFont(new Font(48));
+            textFlow.getChildren().clear();
```

Comment:
is this clear() a necessary part of the test?
setAll() "Clears the ObservableList and adds all the elements..." according to its javadoc

this comment applies here and in all other places (line 164, etc.)

Reply:
Yes clear is not required. Removed from all instances.